### PR TITLE
[cli] Improve interactive scrolling

### LINKED ISSUE #12295: Terminal fast scroll (⏫/⏬) when showing long list of resources.
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

([context](https://pulumi.slack.com/archives/C02970CDD27/p1677532690638549))

Currently, when an update is in progress and there's a long list of resources, a user may use their `⬆` and `⬇` keys to scroll the resources. However, the scroll happens only one line at a time while is not very convenient when the list of resources is long (see screen capture below).

To ease with this, it would be great to be able to scroll one page at a time (instead of one line at a time) using some keyboard shortcuts. Either `J` and `K` or `page up` and `page down` or `CTRL+⬆` and `CTRL+⬇`

image

### Affected area/feature

* pulumi cli

### LINKED ISSUE #12320: Restore previous scroll behavior
A recent CLI update appears to have changed the way scrolling works. Now, when there are more resources than can be shown on a given terminal window, we show a little "more" indicator in the lower-right corner:

image

There are a couple of problems with this new behavior, IMO:

* It's really easy to miss that indicator (I missed it myself until someone pointed it out) and even when you do see it, it's not clear how to interact with it.
* Because the view doesn't scroll automatically as new resources are discovered, it can make for an update experience that looks "stuck" -- e.g., when the visible list of resources is all "created", those still "creating" are offscreen, and the mouse doesn't allow you to scroll down like it used to.
* For those used to scrolling in the terminal with a mouse (like me), having to scroll Pulumi output with the arrow keys (when everything else still works with the mouse) is tedious, especially during a long update or one involving dozens or hundreds of resources.

I'm sure these changes were made for good reason, but the experience feels significantly less usable to me now. Previously, I could scroll up and down with the mouse during an update and see all that I needed to see, so since I don't seem to be able to see more now than I could before, it isn't clear (to me at least) how the new behavior is "better" -- but again, I might just be missing it. 

If others agree, though, I'd strongly suggest we consider returning to the previous behavior, and perhaps make the new behavior optional or configurable somehow.

### PR DESCRIPTION
**Pull request by pgavlin**
These changes make two improvements to scrolling in the interactive display:

1. The interactive display can now be scrolled using the page-down and
   page-up keys. These keys scroll the display N lines at a time, where N is
   the height of the terminal.
2. If the display is scrolled to the bottom, it will now autoscroll it as new lines
   are added to the output.

Fixes #12320 
Fixes #12295

### CODE REVIEW

**Review by dixler — CHANGES_REQUESTED** [review]
Hi. I'm in the process of working on https://github.com/pulumi/pulumi/issues/12320

I would like to block changes to this area while I'm figuring out how to resolve the p1.

**Comment by pgavlin** [comment]
> Hi. I'm in the process of working on #12320
> 
> I would like to block changes to this area while I'm figuring out how to resolve the p1.

Ah, I wasn't aware we were tracking with a bug.

I believe that these changes help address that P1: they allow for faster scrolling and help ensure that newer resources appear on screen in a fashion similar to the earlier renderer.

**Comment by dixler** [comment]
> > Hi. I'm in the process of working on #12320
> > I would like to block changes to this area while I'm figuring out how to resolve the p1.
> 
> Ah, I wasn't aware we were tracking with a bug.
> 
> I believe that these changes help address that P1: they allow for faster scrolling and help ensure that newer resources appear on screen in a fashion similar to the earlier renderer.

I'll take a look.

File: pkg/backend/display/internal/terminal/term.go [hunk 1]
```diff
@@ -154,64 +154,124 @@ func (t *terminal) ShowCursor() {
...
+		d.intermediate = append(d.intermediate, b)
+		return d.stateEscapeIntermediate
 	}
-	var stateParameter stateFunc
-	stateParameter = func(b byte) (stateFunc, string) {
-		if b >= 0x30 && b < 0x40 {
-			return stateParameter, ""
-		}
-		return stateIntermediate(b)
+	if b >= 0x30 && b < 0x7f {
```
Comment (dixler) [inline pkg/backend/display/internal/terminal/term.go:200]: > would appreciate a comment explaining the significance of these values

here as well

**Review by dixler — APPROVED** [review]
Thanks for the PR. I have a few nits, but looks fine as it's an overall improvement.

Scrolling doesn't view prior resources, but that's fine as we may rethink the overall UI experience.

**Comment by dixler** [comment]
It looks like a test is failed. I'll rerun
```
=== FAIL: backend/display TestProgressEvents/template-body.jsoninteractive/80x24/raw (1.35s)
```

**Comment by pgavlin** [comment]
bors merge

**Pull request merged**